File: pypp_cli/test_dir/python/src/operations.py
from pypp_python import int_pow, Ref


def operations_fn():
    print("OPERATIONS RESULTS:")
    # float power
    a: float = 2**3
    print(f"float power: {a}")
    # int power. Note: for safety, you need to use for integer exponentiation because
    #  C++ does not have safe integer exponentiation.
    b: int = int_pow(2, 3)
    print(f"int power: {b}")
    # mod
    c: int = 10 % 3
    print(f"mod: {c}")
    # left shift
    d: int = 1 << 2
    print(f"left shift: {d}")
    # right shift
    e: int = 4 >> 1
    print(f"right shift: {e}")
    # bitwise or
    f: int = 1 | 2
    print(f"bitwise or: {f}")
    # bitwise xor
    g: int = 1 ^ 2
    print(f"bitwise xor: {g}")
    # bitwise and
    h: int = 1 & 2
    print(f"bitwise and: {h}")
    # not
    i: bool = not True
    print(f"not: {i}")
    # usub
    j: int = -5
    print(f"usub: {j}")
    # uadd
    k: int = +5
    print(f"uadd: {k}")
    # invert
    l1: int = ~5
    print(f"invert: {l1}")
    # is
    t1: list[int] = [1, 2]
    t2: Ref(list[int]) = t1
    m: bool = t1 is t2
    print(f"is: {m}")
    # is not
    n: bool = t1 is not t2
    print(f"is not: {n}")
    # in list
    o: bool = 5 in [1, 5]
    print(f"in: {o}")
    # in set
    p: bool = 5 in {1, 5}
    print(f"in set: {p}")
    # in dict
    # Note: I create a test_dict because you can't inline the dicts like the others in
    # Py++. Is this still true or is that possible now?
    test_dict: dict[int, str] = {1: "a", 5: "b"}
    q: bool = 5 in test_dict
    print(f"in dict: {q}")
    # in tuple
    r: bool = 5 in (1, 5)
    print(f"in tuple: {r}")
    # in string
    s: bool = "abc" in "zzabczz"
    print(f"in string: {s}")
    # not in list
    t: bool = 5 not in [1, 2]
    print(f"not in: {t}")
    # not in set
    u: bool = 5 not in {1, 2}
    print(f"not in set: {u}")
    # not in dict
    v: bool = 5 not in test_dict
    print(f"not in dict: {v}")
    # not in tuple
    w: bool = 5 not in (1, 2)
    print(f"not in tuple: {w}")
    # not in string
    x: bool = "abc" not in "zzxyz"
    print(f"not in string: {x}")
    print("printing reference:")
    print(Ref(x))

File: pypp_cli/test_dir/cpp/src/operations.cpp
#include "operations.h"
#include "py_dict.h"
#include "py_list.h"
#include "py_set.h"
#include "py_str.h"
#include "py_tuple.h"
#include "pypp_util/math.h"
#include "pypp_util/print.h"
#include <cmath>

namespace me {
void operations_fn() {
    pypp::print(pypp::PyStr("OPERATIONS RESULTS:"));
    double a = std::pow(2, 3);
    pypp::print(pypp::PyStr(std::format("float power: {}", a)));
    int b = pypp::int_pow(2, 3);
    pypp::print(pypp::PyStr(std::format("int power: {}", b)));
    int c = 10 % 3;
    pypp::print(pypp::PyStr(std::format("mod: {}", c)));
    int d = 1 << 2;
    pypp::print(pypp::PyStr(std::format("left shift: {}", d)));
    int e = 4 >> 1;
    pypp::print(pypp::PyStr(std::format("right shift: {}", e)));
    int f = 1 | 2;
    pypp::print(pypp::PyStr(std::format("bitwise or: {}", f)));
    int g = 1 ^ 2;
    pypp::print(pypp::PyStr(std::format("bitwise xor: {}", g)));
    int h = 1 & 2;
    pypp::print(pypp::PyStr(std::format("bitwise and: {}", h)));
    bool i = !true;
    pypp::print(pypp::PyStr(std::format("not: {}", i)));
    int j = -5;
    pypp::print(pypp::PyStr(std::format("usub: {}", j)));
    int k = +5;
    pypp::print(pypp::PyStr(std::format("uadd: {}", k)));
    int l1 = ~5;
    pypp::print(pypp::PyStr(std::format("invert: {}", l1)));
    pypp::PyList<int> t1({1, 2});
    pypp::PyList<int> &t2 = t1;
    bool m = &t1 == &t2;
    pypp::print(pypp::PyStr(std::format("same objects: {}", m)));
    bool n = &t1 != &t2;
    pypp::print(pypp::PyStr(std::format("different objects: {}", n)));
    bool o = pypp::PyList({1, 5}).contains(5);
    pypp::print(pypp::PyStr(std::format("in: {}", o)));
    bool p = pypp::PySet({1, 5}).contains(5);
    pypp::print(pypp::PyStr(std::format("in set: {}", p)));
    pypp::PyDict<int, pypp::PyStr> test_dict = {{1, pypp::PyStr("a")},
                                                {5, pypp::PyStr("b")}};
    bool q = test_dict.contains(5);
    pypp::print(pypp::PyStr(std::format("in dict: {}", q)));
    bool r = pypp::PyTup(1, 5).contains(5);
    pypp::print(pypp::PyStr(std::format("in tuple: {}", r)));
    bool s = pypp::PyStr("zzabczz").contains(pypp::PyStr("abc"));
    pypp::print(pypp::PyStr(std::format("in string: {}", s)));
    bool t = !pypp::PyList({1, 2}).contains(5);
    pypp::print(pypp::PyStr(std::format("not in: {}", t)));
    bool u = !pypp::PySet({1, 2}).contains(5);
    pypp::print(pypp::PyStr(std::format("not in set: {}", u)));
    bool v = !test_dict.contains(5);
    pypp::print(pypp::PyStr(std::format("not in dict: {}", v)));
    bool w = !pypp::PyTup(1, 2).contains(5);
    pypp::print(pypp::PyStr(std::format("not in tuple: {}", w)));
    bool x = !pypp::PyStr("zzxyz").contains(pypp::PyStr("abc"));
    pypp::print(pypp::PyStr(std::format("not in string: {}", x)));
    pypp::print(pypp::PyStr("printing reference:"));
    pypp::print(&x);
}

} // namespace me